docs/reference/apply_and_sync.md

```diff
@@ -0,0 +1,88 @@
+
+This dance is necessary to give users in IaSQL the ability to migrate from *any* cloud state to the state they have defined in their database. This is IaSQL's superpower: bringing a greater reliability to producing the cloud state the developer specifies, so they can worry about more important things.
+
+## Mapper Footguns
+
+This section is meant for anyone currently developing in IaSQL, so it will change much more frequently compared to the sections above.
+
+* TypeORM sometimes violates the entity types you have provided, especially around nullable types, where the entity may be something like `foo?: string` which implies `string | undefined` but TypeORM will treat it as `string | null`. This can trip you up in the `equals` function as `Object.is` distinguishes between `undefined` and `null` so you will run into infinite `update` loops where it is detecting a difference but there is no actionable change to be had. (Similarly sometimes arrays are actually unordered sets, so comparing array element-by-element is incorrect and you need to check each element is inside of the other array, regardless of order, and some strings are actually JSON or other human-editable data formats and non-significant-whitespace differences can also cause `update` loops.
+* Deeply-nested entities get translated into many tables in TypeORM, and unique constraints on those tables can cause lots of pain with the entities that were generated from the cloud APIs, as they don't have the the database-only columns that let TypeORM know when the entity already exists in the database. Unfortunately for now you have to manually find these duplicates and then patch your sub-entity with these DB-only fields or the DB CRUD operations will blow up in your face.
+* Picking the right field for the `@cloudId` in order to both: 1) properly distinguish records in the cloud and 2) still support distinguishing records in the database can sometimes be a difficult job, particularly if the only real candidate is a cloud-generated property.
+* Records that are actually really read-only from the cloud and can't actually be mutated by end users (but are still useful to have to make inserts/updates of other entities safer via join logic) need to have weirdly "backwards" cloud CRUD functions that grab the cloud cache record and re-save it to the database (or delete the user-inserted entity outright).
+* The create-only/create-only-with-unique-columns entities currently require 100% type-unsafe actions to manipulate the DB cache. There are other sections of the codebase where type safety / guarding by the compiler is broken, but this is the worst/most-dangerous one.
```

Comment:
what do you mean with `The create-only/create-only-with-unique-columns entities`? maybe an example here would be good

Reply:
Oh, this is referring to the security group mapper and ones like it, where you have to `if(ctx?.memo?.db?.<EntityName>?.<IdHere>) ctx.memo.db.<EntityName>.<IdHere> = myReplacementThing;`

100% not typechecked by Typescript so it's always a bit dangerous to do it without extensive integration testing.

Reply:
ohh I see. Same as before, maybe a link to a place in the code where this is being done for context would be good!

Reply:
```suggestion
* The create-only/create-only-with-unique-columns entities currently require [100% type-unsafe actions to manipulate the DB cache](https://github.com/iasql/iasql-engine/blob/main/src/modules/0.0.15/aws_security_group/index.ts#L171). There are other sections of the codebase where type safety / guarding by the compiler is broken, but this is the worst/most-dangerous one.
```